src/base/detail/AlgorithmsImpl.hh

```diff
@@ -0,0 +1,305 @@
+//----------------------------------*-C++-*----------------------------------//
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//---------------------------------------------------------------------------//
+/*!
+ * \file AlgorithmsImpl.hh
+ * \brief Algorithm implementations directly derived from LLVM libc++
+ */
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <iterator>
+#include <type_traits>
+#include "base/Macros.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+// Forward declare cuda-compatible swap/move
+template<class T>
+CELER_FUNCTION void
+cswap(T&, T&) noexcept(std::is_nothrow_move_constructible<T>::value&&
+                           std::is_nothrow_move_assignable<T>::value);
+
+template<class T>
+CELER_CONSTEXPR_FUNCTION auto cmove(T&&) noexcept ->
+    typename std::remove_reference<T>::type&&;
+
+namespace detail
+{
+//---------------------------------------------------------------------------//
+template<class RandomAccessIt>
+using difference_type_t =
+    typename std::iterator_traits<RandomAccessIt>::difference_type;
+
+//---------------------------------------------------------------------------//
+// LOWER_BOUND
+//---------------------------------------------------------------------------//
+//!@{
+/*!
+ * Perform division-by-two quickly for positive integers.
+ *
+ * See llvm.org/PR39129.
+ */
+template<typename Integral>
+CELER_CONSTEXPR_FUNCTION
+    typename std::enable_if<std::is_integral<Integral>::value, Integral>::type
+    half_positive(Integral value)
```

Comment:
Some algorithms start with `c` and some do not start with `c` with the discriminant being (if I understood correctly) "name is already use in `std` namespace".   [If the prefix stays necessary, see oher comment], should't we for clarity/regularity-of-pattern use the prefix for all algorithm (that have the `CELER_..._FUNCTION` decoration)?

Reply:
That's a good point... right now the criteria is "do the std algorithms we use on celeritas data structures (or vice versa) use these functions/algorithms": and that's pretty arbitrary. I don't know offhand if the standard specifies whether *their* internal function usage has to be fully qualified, or which functions besides swap are supposed to be unqualified so they can use ADL.